astropy/coordinates/builtin_frames/galactocentric.py

```diff
@@ -85,80 +95,134 @@ class galactocentric_frame_defaults(ScienceState):
+                Galactocentric solar parameters
+            - "references" : Dict[str, Union[str, Sequence[str]]]
+                References for "value".
+                Fields are str or sequence of str.
+
+        Raises
+        ------
+        KeyError
+            If invalid string input to registry
+            to retrieve solar parameters for Galactocentric frame.
+
+        """
+        # Resolve the meaning of 'latest': latest parameter set is from v4.0
         # - update this as newer parameter choices are added
         if arg == 'latest':
             arg = cls._latest_value
 
-        params = dict()
-        references = dict()
-
-        # Currently, all versions use the same sky position for Sgr A*:
-        params['galcen_coord'] = ICRS(ra=266.4051*u.degree,
-                                      dec=-28.936175*u.degree)
-        references['galcen_coord'] = \
-            'http://adsabs.harvard.edu/abs/2004ApJ...616..872R'
-
-        # The roll angle is the same for both frames:
-        params['roll'] = 0 * u.deg
-
-        if arg == 'pre-v4.0':
-            params['galcen_distance'] = 8.3 * u.kpc
-            references['galcen_distance'] = \
-                'https://ui.adsabs.harvard.edu/#abs/2009ApJ...692.1075G'
-
-            params['galcen_v_sun'] = r.CartesianDifferential([11.1,
-                                                              220+12.24,
-                                                              7.25]*u.km/u.s)
-            references['galcen_v_sun'] = \
-                ['https://ui.adsabs.harvard.edu/#abs/2010MNRAS.403.1829S',
-                 'https://ui.adsabs.harvard.edu/#abs/2015ApJS..216...29B']
-
-            params['z_sun'] = 27.0 * u.pc
-            references['z_sun'] = \
-                'https://ui.adsabs.harvard.edu/#abs/2001ApJ...553..184C'
-
-        elif arg == 'v4.0':
-            params['galcen_distance'] = 8.122 * u.kpc
-            references['galcen_distance'] = \
-                'https://ui.adsabs.harvard.edu/abs/2018A%26A...615L..15G'
-
-            params['galcen_v_sun'] = r.CartesianDifferential([12.9,
-                                                              245.6,
-                                                              7.78]*u.km/u.s)
-            references['galcen_v_sun'] = \
-                ['https://ui.adsabs.harvard.edu/abs/2018RNAAS...2..210D',
-                 'https://ui.adsabs.harvard.edu/abs/2018A%26A...615L..15G',
-                 'https://ui.adsabs.harvard.edu/abs/2004ApJ...616..872R']
-
-            params['z_sun'] = 20.8 * u.pc
-            references['z_sun'] = \
-                'https://ui.adsabs.harvard.edu/abs/2019MNRAS.482.1417B'
+        # Get the state from the registry.
+        # Copy to ensure registry is immutable to modifications of "_value".
+        # Raises KeyError if `arg` is invalid string input to registry
+        # to retrieve solar parameters for Galactocentric frame.
+        state = cls._registry[arg].copy()
```

Comment:
I believe `dict.copy()` is a shallow copy, so maybe we want `copy.deepcopy(cls._registry[arg])` to be safe? For current parameter classes, I guess it doesn't matter though..

Reply:
good point. I'll change it.

Reply:
I agree a deep copy is necessary. Unfortunately deep copy does not work on a MappingProxyType, but I have found a fix for this, which is useful here, but maybe also elsewhere in Astropy

```
from types import MappingProxyType
import copyreg, copy, pickle

def pickle_mappingproxytype(mpt):
    return dict, (tuple(mpt.items()), )

copyreg.pickle(MappingProxyType, pickle_mappingproxytype)
```

As example
```
d = {"value": MappingProxyType({"A": 1}), }

c = copy.deepcopy(d)
c["value"]["A"] = 2

print(c)
print(d)
```
> {'value': {'A': 2}}
> {'value': mappingproxy({'A': 1})}

Reply:
How about not making a copy at all, and leaving it up to the user to be careful? It would seem fairly standard behaviour (and the `MappingProxy` avoids the most obvious errors).

Reply:
In the current `galactocentric_frame_defaults`, when the `_value` is changed this doesn't change the hardcoded registry. I was trying to keep that expectation. The problem with making a shallow copy or no copy is the mutability of the registry. A different fix I looked at was to MappingProxy the whole state when set from the registry. Unfortunately this only makes the top-level read-only, so nested dictionaries, like "parameters" and "references" are mutable. This can approximately be fixed by wrapping both of these in MappingProxy, but for any custom metadata this is not a solution. I think that `deepcopy` is the only easily implemented means of maintaining separation between the state and the registry.

Thankfully, the `copyreg` solution uses only built-in libraries. The only downside is that registering a serialization scheme for MappingProxy here might work in all of Astropy. This isn't actually a problem as 1) MappingProxy is used nowhere else in Astropy and 2) it might even be a desirable outcome to be able to deepcopy a MappingProxy and get a dictionary, not an error. 

This code snippet should be somewhere in `astropy.utils`, not here, but I don't see an obvious place. Thoughts?